internal/terraform/node_module_expand.go

```diff
@@ -98,7 +107,7 @@ func (n *nodeExpandModule) DependsOn() []*addrs.Reference {
 
 // GraphNodeReferenceOutside
 func (n *nodeExpandModule) ReferenceOutside() (selfPath, referencePath addrs.Module) {
-	return n.Addr, n.Addr.Parent()
+	return n.Addr.Parent(), n.Addr.Parent()
```

Comment:
OK, this is starting to make more sense. The original graph shape was structured around the variables and outputs, so only `referencePath` was really being used to connect `for_each` references. There was no real reference directly to a module until expansion was introduced, and even then it relied on the output objects for coordination. I think we may have a bug open somewhere about a module with no outputs not being referencable which this might also fix.